tests/models/test_fields.py

```diff
@@ -0,0 +1,202 @@
+
+    def test_field_knows_its_name(self):
+        instance = create_instance(Field())
+        self.assertEqual('attr', instance.__class__.attr._name)
+
+    def test_field_has_none_as_default(self):
+        instance = create_instance(Field())
+        self.assertIsNone(instance.attr)
+
+    def test_field_does_not_store_default_in_dict(self):
+        instance = create_instance(Field())
+        self.assertNotIn('attr', instance.__dict__)
+
+    def test_field_assigment_and_retrival(self):
+        instance = create_instance(Field())
+        instance.attr = 1234
+        self.assertEqual(1234, instance.attr)
+        self.assertEqual(1234, instance.__dict__['attr'])
+
+    def test_field_can_be_reassigned(self):
+        instance = create_instance(Field())
+        instance.attr = 1234
+        instance.attr = 5678
+        self.assertEqual(5678, instance.attr)
+
+    def test_field_can_be_deleted(self):
+        instance = create_instance(Field())
+        instance.attr = 1234
+        del instance.attr
+        self.assertEqual(None, instance.attr)
+        self.assertNotIn('attr', instance.__dict__)
+
+    def test_field_can_be_set_to_none(self):
+        instance = create_instance(Field())
+        instance.attr = 1234
+        instance.attr = None
+        self.assertEqual(None, instance.attr)
+        self.assertNotIn('attr', instance.__dict__)
+
+    def test_field_can_be_set_default(self):
+        default = object()
+        instance = create_instance(Field(default=default))
+        instance.attr = 1234
+        instance.attr = default
+        self.assertEqual(default, instance.attr)
+        self.assertNotIn('attr', instance.__dict__)
+
+
+class FieldTest(unittest.TestCase):
+    def test_default_handling(self):
+        instance = create_instance(Field(default=1234))
+        self.assertEqual(1234, instance.attr)
+
+    def test_type_checking(self):
+        instance = create_instance(Field(type=set))
+        instance.attr = set()
+
+        with self.assertRaises(TypeError):
+            instance.attr = 1234
+
+    def test_choices_checking(self):
+        instance = create_instance(Field(choices=(1, 2, 3)))
+        instance.attr = 1
+
+        with self.assertRaises(TypeError):
+            instance.attr = 4
+
+    def test_default_respects_type_check(self):
+        with self.assertRaises(TypeError):
+            create_instance(Field(type=int, default='123'))
+
+    def test_default_respects_choices_check(self):
+        with self.assertRaises(TypeError):
+            create_instance(Field(choices=(1, 2, 3), default=5))
+
+
+class StringTest(unittest.TestCase):
+    def test_default_handling(self):
+        instance = create_instance(String(default='abc'))
+        self.assertEqual('abc', instance.attr)
+
+    def test_str_allowed(self):
+        instance = create_instance(String())
+        instance.attr = str('abc')
+        self.assertEqual(b'abc', instance.attr)
```

Comment:
For increased Python 3 compat, rename this test to `test_native_str_allowed` and remove the `b` prefix.

Then add another test for `test_bytes_allowed` with `b'foo'` syntax in both assignment and assert.